src/pybind/mgr/deepsea/module.py

```diff
@@ -307,13 +318,42 @@ def _set_last_failure_msg(self, msg):
     # Note: this is pretty braindead and doesn't implement the full eventsource
     # spec, but it *does* implement enough for us to listen to events from salt
     # and potentially do something with them.
-    # TODO: How are we going to deal with salt-api dying, or mgr failing over,
-    #       or other unforeseen glitches when we're waiting for particular jobs
-    #       to complete?  What's to stop things falling through the cracks?
-    #       Do completions we're waiting on need to be persisted somewhere?
     def _read_sse(self):
         event = {}
         try:
+            # Just starting the event reader; if we've made it here, we know we're
+            # talking to salt-api (_do_request would have raised an exception if the
+            # response wasn't ok), so check if there's any completions inflight that
+            # need to be dealt with.  This handles the case where some command was
+            # invoked, then salt-api died somehow, and we reconneced, but missed the
+            # completion at the time it actually happened.
+            for tag in list(self._all_completions):
+                self.log.info("Found event {} inflight".format(tag))
+                resp = self._do_request_with_login("POST", data = {
+                    "client": "runner",
+                    "fun": "jobs.lookup_jid",
+                    "jid": tag.split('/')[2]
+                })
+                if resp.ok:
+                    # jobs.lookup_jid returns a dict keyed by hostname.
+                    return_dict = resp.json()['return'][0]
+                    if return_dict:
+                        # If the job is complete, there'll be one item in the dict.
+                        self.log.info("Event {} complete".format(tag))
+                        # The key is the salt master hostname, but we don't care
+                        # about that, so just grab the data.
+                        data = next(iter(return_dict.items()))[1]
+                        self._all_completions[tag]._process_result(data)
+                        # TODO: decide whether it's bad to delete the completion
+                        # here -- would we ever need to resurrect it?
+                        del self._all_completions[tag]
+                    else:
+                        # if the job is not complete, there'll be nothing in the dict
+                        self.log.info("Event {} still pending".format(tag))
+                else:
+                    # TODO: What do we do if the request fails?
```

Comment:
Maybe throw an exception and let the user know it?

Reply:
The user won't see this directly, as this is happening internally to the module, not in response to user commands. I've rejigged it slightly so it logs a warning, then can keep iterating through any other outstanding completions, then carry on listening to the salt event bus.

Reply:
We have multiple ways no notify users: one way would be to use the Completion objects to transfer the error, like in `rook.module.RookOrchestrator#wait`. Another one would be to leverage the `progress` module and set the `message` to `str(e)`

Reply:
If this request fails, all it means is that we weren't able to query the status of the inflight job. It doesn't tell us if the job succeeded or failed, just that we currently have NFI what's going on with that job.